blacksheep/testing/simulator.py

```diff
@@ -0,0 +1,97 @@
+import abc
+from typing import Dict, Optional
+from urllib import parse
+
+from blacksheep.contents import Content
+from blacksheep.messages import Request
+from blacksheep.server.application import Application
+from blacksheep.server.responses import Response
+from blacksheep.testing.helpers import get_example_scope
+
+
+def _create_scope(
+    method: str,
+    path: str,
+    headers: Optional[Dict[str, str]] = None,
+    query: Optional[Dict[str, str]] = None,
+) -> Dict:
+    """Creates a mocked ASGI scope"""
+    if headers is not None:
+        headers = [(key.encode(), value.encode()) for key, value in headers.items()]
+
+    query = parse.urlencode(query).encode() if query else b""
+
+    scope = get_example_scope(method, path, extra_headers=headers, query=query)
+    return scope
+
+
+class AbstractTestSimulator:
+    """An abstract class for custom Test simulator clients"""
+
+    @abc.abstractmethod
+    async def send_request(
+        self,
+        method: str,
+        path: str,
+        headers: Optional[Dict[str, str]] = None,
+        query: Optional[Dict[str, str]] = None,
+        content: Optional[Content] = None,
+    ):
+        """Entrypoint for all HTTP methods
+
+        The method is main entrypoint for all TestClient methods
+            - get
+            - post
+            - patch
+            - put
+            - delete
+        Then you can define an own TestClient, with the custom logic.
+        """
+        pass
```

Comment:
This `pass` statement can be removed, since the abstract method declaration has a docstring. This is also useful for code coverage report, since it gives a false negative for `pass`:

![image](https://user-images.githubusercontent.com/2576032/120888226-74acae80-c5f7-11eb-8b0c-20df78425f45.png)